docs/modules/ROOT/pages/design/cloudevents.adoc

```diff
@@ -0,0 +1,125 @@
+= Cloud Event Delivery Pipeline
+== Overview
+
+KubeArchive uses Cloud Events that are generated by a Knative Eventing ApiServerSource to discover when resources on
+the kubernetes cluster have changed. These Cloud Events contain the json representation of the Kubernetes resource that
+has changed. The KubeArchive Sink (sink) receives these messages and determines, using CEL expression from the
+KubeArchiveConfig, if the data contained in the Cloud Event should be written to the database and potentially deleted
+from the cluster. While the sink is processing a Cloud Event, it is possible at many points for an error to occur,
+especially when the sink has to make a request of the network, particularly, to the Kubernetes API or the database. This
+page proposes how KubeArchive set up its Cloud Event delivery pipeline and how it should consume Cloud Events to achieve
+the following goals:
+
+* KubeArchive receives all Cloud Events from the Knative ApiServerSource
+* Cloud Events sent to KubeArchive can be replayed
+* Cloud Events received by KubeArchive, that cannot be processed successfully, are reported back to the sender as
+  delivery failure
+* Cloud Events sent to the sink will attempt to be redelivered if the initial delivery fails
+* Cloud Events that could not be delivered are replayed to KubeArchive automatically when KubeArchive becomes available
+  again
+
+== Design
+
+KubeArchive needs to use the functionality provided by Knative Eventing, Brokers and Triggers as well as features
+provided by the Cloud Events Specification to achieve an event driven architecture, where events are not lost and events
+can be replayed safely without worries about side effects.
+
+This diagram provides a high level overview of the life cycle of Cloud Events look like in KubeArchive and how the
+different components of the Cloud Event Delivery Pipeline interact with eachother.
+
+image::ce-pipeline.png[]
+
+=== Knative Eventing Brokers
+
+Instead of the ApiServerSource sending Cloud Events directly to the sink, the Cloud Events should instead be sent to a
+Knative Eventing Broker. Brokers can receive, persist, and send messages. Using the broker allows messages to be
+sent somewhere that they can be stored long term and is separate from the KubeArchive sink and Knative Eventing
+ApiServerSource, so if either are restarted, the Cloud Events won't be lost. KubeArchive will use two brokers. One
+primary broker (called broker) and secondary broker (called dead letter sink). The default configuration of
+Knative Eventing will have these brokers be a created as In Memory Brokers, but changing the Knative Eventing
+configuration can allow this broker to be created, for example, as a Kafka Broker instead
```

Comment:
```suggestion
Instead of sending directly events directly to KubeArchive's sink, we decided to send the events to a Knative Eventing Broker. These can persist events in case of consumer or producer failure. By default KubeArchive uses two in-memory brokers: primary and dead letter sink. The primary broker receives the events from the ApiServerSource and sends them to the dead letter sink broker if its unable to deliver them to KubeArchive.
```